File: 01. Creacionales/02. Factory/01. factory.js
/***

Imaginemos que estamos construyendo una aplicación que necesita conectarse a diferentes bases de datos, 
como MySQL, PostgreSQL y MongoDB. Queremos tener un código limpio y modular que nos permita crear fácilmente 
las conexiones a estas bases de datos sin tener que preocuparnos por los detalles de implementación de cada una.

Aquí es donde el patrón Factory puede ser útil. Podemos utilizar el patrón Factory para encapsular la lógica 
de creación de las conexiones a las bases de datos y proporcionar una interfaz común para interactuar con ellas.

*/

// Comencemos definiendo una interfaz común para nuestras conexiones a bases de datos:

class DatabaseConnection {
    connect() {
        throw new Error("Método connect() debe ser implementado por las subclases.");
    }

    query(sql) {
        throw new Error("Método query() debe ser implementado por las subclases.");
    }

    disconnect() {
        throw new Error("Método disconnect() debe ser implementado por las subclases.");
    }
}

// A continuación, definimos nuestras clases concretas que implementan la interfaz DatabaseConnection para cada base de datos:

class MySQLConnection extends DatabaseConnection {
    connect() {
        console.log("Conectando a la base de datos MySQL...");
    }

    query(sql) {
        console.log(`Ejecutando la consulta SQL en la base de datos MySQL: ${sql}`);
    }

    disconnect() {
        console.log("Desconectando de la base de datos MySQL...");
    }
}

class PostgreSQLConnection extends DatabaseConnection {
    connect() {
        console.log("Conectando a la base de datos PostgreSQL...");
    }

    query(sql) {
        console.log(`Ejecutando la consulta SQL en la base de datos PostgreSQL: ${sql}`);
    }

    disconnect() {
        console.log("Desconectando de la base de datos PostgreSQL...");
    }
}

class MongoDBConnection extends DatabaseConnection {
    connect() {
        console.log("Conectando a la base de datos MongoDB...");
    }

    query(sql) {
        console.log(`Ejecutando la consulta SQL en la base de datos MongoDB: ${sql}`);
    }

    disconnect() {
        console.log("Desconectando de la base de datos MongoDB...");
    }
}

// Luego, definimos nuestro Factory que se encargará de crear las conexiones a las bases de datos:

class DatabaseConnectionFactory {
    static createConnection(databaseType) {
        switch (databaseType) {
            case "MySQL":
                return new MySQLConnection();
            case "PostgreSQL":
                return new PostgreSQLConnection();
            case "MongoDB":
                return new MongoDBConnection();
            default:
                throw new Error(`Tipo de base de datos no soportado: ${databaseType}`);
        }
    }
}

// Finalmente, podemos utilizar el patrón Factory para crear conexiones a diferentes bases de datos:

const mysqlConnection = DatabaseConnectionFactory.createConnection("MySQL");
mysqlConnection.connect();
mysqlConnection.query("SELECT * FROM users");
mysqlConnection.disconnect();

const postgresqlConnection = DatabaseConnectionFactory.createConnection("PostgreSQL");
postgresqlConnection.connect();
postgresqlConnection.query("SELECT * FROM users");
postgresqlConnection.disconnect();

const mongodbConnection = DatabaseConnectionFactory.createConnection("MongoDB");
mongodbConnection.connect();
mongodbConnection.query("db.users.find()");
mongodbConnection.disconnect();

File: 01. Creacionales/02. Factory/01. factory.py
class DatabaseConnection:
    def connect(self):
        raise NotImplementedError("El método connect() debe ser implementado por las subclases.")

    def query(self, sql):
        raise NotImplementedError("El método query() debe ser implementado por las subclases.")

    def disconnect(self):
        raise NotImplementedError("El método disconnect() debe ser implementado por las subclases.")


class MySQLConnection(DatabaseConnection):
    def connect(self):
        print("Conectando a la base de datos MySQL...")

    def query(self, sql):
        print(f"Ejecutando la consulta SQL en la base de datos MySQL: {sql}")

    def disconnect(self):
        print("Desconectando de la base de datos MySQL...")


class PostgreSQLConnection(DatabaseConnection):
    def connect(self):
        print("Conectando a la base de datos PostgreSQL...")

    def query(self, sql):
        print(f"Ejecutando la consulta SQL en la base de datos PostgreSQL: {sql}")

    def disconnect(self):
        print("Desconectando de la base de datos PostgreSQL...")


class MongoDBConnection(DatabaseConnection):
    def connect(self):
        print("Conectando a la base de datos MongoDB...")

    def query(self, sql):
        print(f"Ejecutando la consulta SQL en la base de datos MongoDB: {sql}")

    def disconnect(self):
        print("Desconectando de la base de datos MongoDB...")


class DatabaseConnectionFactory:
    @staticmethod
    def create_connection(database_type):
        if database_type == "MySQL":
            return MySQLConnection()
        elif database_type == "PostgreSQL":
            return PostgreSQLConnection()
        elif database_type == "MongoDB":
            return MongoDBConnection()
        else:
            raise ValueError(f"Tipo de base de datos no soportado: {database_type}")


mysql_connection = DatabaseConnectionFactory.create_connection("MySQL")
mysql_connection.connect()
mysql_connection.query("SELECT * FROM users")
mysql_connection.disconnect()

postgresql_connection = DatabaseConnectionFactory.create_connection("PostgreSQL")
postgresql_connection.connect()
postgresql_connection.query("SELECT * FROM users")
postgresql_connection.disconnect()

mongodb_connection = DatabaseConnectionFactory.create_connection("MongoDB")
mongodb_connection.connect()
mongodb_connection.query("db.users.find()")
mongodb_connection.disconnect()